test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.loopopts;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @bug 8323220
+ * @summary Test loop invariant code motion for cmp nodes through reassociation
+ * @library /test/lib /
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
+ */
+public class InvariantCodeMotionReassociateCmp {
+    private static final int size = 500;
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @DontInline
+    private void blackhole() {}
+
+    @Test
+    @Arguments({Argument.RANDOM_EACH, Argument.RANDOM_EACH})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public int[] equalsAddInt(int inv1, int inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
```

Comment:
If you return instead of break, do you still need the `blackhole`?

Reply:
No we don't need it. But it would feel strange to have a loop with nothing in it other than comparisons of an induction variable.